posts/2017-05-02-io-monad-for-cats.md

```diff
@@ -0,0 +1,404 @@
+
+`Sync` simply describes the `IO.apply` function (in the typeclasses, this function is called `delay`).  Which is to say, any type constructor `F[_]` which has a `Sync[F]` has the capability to suspend *synchronous* side-effecting code.  `Async` is very similar to this in that it describes the `async` function.  So any type constructor `F[_]` which has an `Async[F]` can suspend *asynchronous* side-effecting code.  `LiftIO` should be familiar to Haskell veterans, and is broadly useful for defining parametric signatures and composing monad transformer stacks.
+
+`Effect` is where everything is brought together.  In addition to being able to suspend synchronous and asynchronous side-effecting code, anything that has an `Effect` instance may also be *asynchronously interpreted* into an `IO`.  The way this is specified is using the `runAsync` function:
+
+```tut:silent
+import cats.effect.{Async, LiftIO, Sync}
+
+trait Effect[F[_]] extends Sync[F] with Async[F] with LiftIO[F] {
+  def runAsync[A](fa: F[A])(cb: Either[Throwable, A] => IO[Unit]): IO[Unit]
+}
+```
+
+<!--    please don't remove; my editor is glitchy :-(
+```
+-->
+
+What this is saying is that any `Effect` must define the ability to evaluate as a side-effect, but of course, we don't want to have side-effects in our pure and reasonable code.  So how are side-effects *purely* represented?  With `IO`!
+
+From a parametric reasoning standpoint, `IO` means "here be effects", and so any type signature which involves `IO` thus also involves side-effects (well, *effects* anyway), and any type signature which requires side-effects must also involve `IO`.  This bit of trickery allows us to reason about `Effect` in a way that would have been much harder if we had defined `unsafeRunAsync` as a member, and it ensures that downstream projects which write code abstracting over `Effect` types can do so without using any `unsafe` functions if they so choose (especially when taken together with the `liftIO` function).
+
+## Conclusion
+
+The lack of a production-ready `Task`-like type fully integrated into the cats ecosystem has been a sticking point for a lot of people considering adopting cats.  With the introduction of [cats-effect](https://github.com/typelevel/cats-effect), this should no longer be a problem!  As of right now, the only releases are snapshots with hash-based versions, the latest of which can be found in the maven badge at the top of the readme.  These snapshots are stable versions (in the repeatable-build sense), but they should not be considered stable, production-ready, future-proof software.  We are quickly moving towards a final 0.1 release, which will depend on cats-core and will represent the stable, finalized API.
```

Comment:
For `tut` I've used the hash-based release, which I think is "good enough" for a blog post.